### PR TITLE
[zero] update sharded optim v2

### PR DESCRIPTION
**Pull request by ver217**
1. Update sharded optim v2 with new sharded model v2
2. Fix some bugs about wrong grad shape when dp size = 1

### CODE REVIEW

File: colossalai/zero/sharded_optim/sharded_optim_v2.py
Comment (FrankLeeeee) [inline colossalai/zero/sharded_optim/sharded_optim_v2.py:72]: `.to(self.device)` is not needed?

Comment (FrankLeeeee) [inline colossalai/zero/sharded_optim/sharded_optim_v2.py:72]: `.to(self.device)` is not needed?

File: colossalai/zero/sharded_optim/sharded_optim_v2.py
Comment (feifeibear) [inline colossalai/zero/sharded_optim/sharded_optim_v2.py:73]: why not merge this if with the above one?

File: colossalai/zero/sharded_optim/sharded_optim_v2.py
Comment (feifeibear) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:72]: I think you do not need to trigger time-consuming communication (in shard and gather). just view the local part of the fp16 payload?

File: colossalai/zero/sharded_optim/sharded_optim_v2.py
Comment (feifeibear) [inline colossalai/zero/sharded_optim/sharded_optim_v2.py:105]: Just copy the valid piece of p.data to p.col_attr and no shard and gather required. It can reduce some time in __init___ but not significant to the overall training performance as the init just executed once.
You can add a TODO for such optimization and we can do it latter.

File: colossalai/zero/sharded_optim/sharded_optim_v2.py
Comment (feifeibear) [inline colossalai/zero/sharded_optim/sharded_optim_v2.py:79]: Where did you shard the momentum and variance data?

File: colossalai/zero/sharded_optim/sharded_optim_v2.py
Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:72]: > `.to(self.device)` is not needed?

I think it influences the device of optimizer states, just in case fp16 param is not on CPU when we want to offload optimizer states.

Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:72]: > I think you do not need to trigger time-consuming communication (in shard and gather). just view the local part of the fp16 payload?

Absolutely I can do this. However, in order to support more shard strategies in the future, I think this code snippet is more general.

File: colossalai/zero/sharded_optim/sharded_optim_v2.py
Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:105]: I will discuss you in the office later.

File: colossalai/zero/sharded_optim/sharded_optim_v2.py
Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:79]: As p.data and p.grad are sharded, optimizer states are sharded naturally.

Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:79]: Pytorch optimizers do a lazy init of optimizer states, whose shape depends on the shape of p.data when doing optimizer.step().

File: colossalai/zero/sharded_optim/sharded_optim_v2.py
Comment (feifeibear) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:73]: now you can just add a TODO(ver217) comments here

File: colossalai/zero/sharded_optim/sharded_optim_v2.py
Comment (feifeibear) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:80]: I see

File: colossalai/zero/sharded_optim/sharded_optim_v2.py
Comment (ver217) [reply colossalai/zero/sharded_optim/sharded_optim_v2.py:73]: Done

File: tests/test_zero_data_parallel/test_sharded_optim_v2.py
Comment (feifeibear) [inline tests/test_zero_data_parallel/test_sharded_optim_v2.py:39]: add bert?

Comment (ver217) [reply tests/test_zero_data_parallel/test_sharded_optim_v2.py:39]: OK

Comment (ver217) [reply tests/test_zero_data_parallel/test_sharded_optim_v2.py]: Done

**Review by feifeibear — APPROVED** [review]
LGTM

**Pull request merged**
